storage/pebble/pebble.go

```
@@ -0,0 +1,283 @@
+	return pdb.db.Delete(append(prefix, key...), pebble.Sync)
+}
+
+func (pdb *PebbleDB) GetMany(prefix []byte, keys [][]byte, iterator func(basedb.Obj) error) error {
+	for _, key := range keys {
+		// Create a new slice for the full key (prefix + key)
+		fullKey := make([]byte, len(prefix)+len(key))
+		copy(fullKey, prefix)
+		copy(fullKey[len(prefix):], key)
+
+		// Retrieve the value from Pebble.
+		value, closer, err := pdb.db.Get(fullKey)
+		if err != nil {
+			// If the key isn't found, skip it.
+			if errors.Is(err, pebble.ErrNotFound) {
+				continue
+			}
+			return err
+		}
+
+		// Since the returned value is only valid until closer.Close(),
+		// we make a copy of it.
+		vcopy := make([]byte, len(value))
+		copy(vcopy, value)
+
+		// Close the closer to release resources.
+		if err := closer.Close(); err != nil {
+			return err
+		}
+
+		// Wrap the key/value in an object satisfying basedb.Obj.
+		obj := basedb.Obj{
+			Key:   key,
+			Value: vcopy,
+		}
+
+		// Call the iterator callback.
+		if err := iterator(obj); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func makePrefixIter(dbOrBatch pebble.Reader, prefix []byte) (*pebble.Iterator, error) {
+	keyUpperBound := func(b []byte) []byte {
+		end := make([]byte, len(b))
+		copy(end, b)
+		for i := len(end) - 1; i >= 0; i-- {
+			end[i] = end[i] + 1
+			if end[i] != 0 {
+				return end[:i+1]
+			}
+		}
+		return nil // no upper-bound
+	}
+
+	prefixIterOptions := func(prefix []byte) *pebble.IterOptions {
+		return &pebble.IterOptions{
+			LowerBound: prefix,
+			UpperBound: keyUpperBound(prefix),
+		}
+	}
+
+	return dbOrBatch.NewIter(prefixIterOptions(prefix))
+}
+
+func (pdb *PebbleDB) GetAll(prefix []byte, iterator func(int, basedb.Obj) error) error {
+	iter, err := makePrefixIter(pdb.db, prefix)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := iter.Close(); err != nil {
+			pdb.logger.Error("close iterator on get all", zap.Error(err))
+		}
+	}()
```

Why not propagate `err` instead of logging it (like you are doing in `FullGC`) ? We probably should handle it the same way for all cases

---

in the other implementation - BadgerDB - _close_ method does not return error hence here I decided to align the functionality to follow same semantics; also, in PebbleDB `Close()` returns the same "accumulated errors" that the `iter.Close()` returns at the end of the function, so the caller receives it either way; the only issue with logging it here is that it would likely be double logged: here, and on the caller side (most likely)